CAPE/CINH using 2-m fields: develop branch

keltonhalbert opened the pull request:
Work in progress for #1273, using the modifications provided by @EdwardColon-NOAA via email.

I currently do not have access to WCOSS2, so I am trying to install dependencies locally in order to confirm that it compiles without error. In the mean-time, I wanted to go ahead and provide the starting point, and ask a few clarifying questions (@BenjaminBlake-NOAA) :

1. In MISCLN.f, I just want to double-check and make sure that the T1D and TBND arrays are potential temperature, and not air temperature. It seems like they should be, based on the code, but I could not find a definitive answer and wanted to make absolutely sure. 

2. Right now, the mixed-layer and surface-based parcel calculations are overwriting the lowest model grid level values. Is that something that we want to do? It may not ultimately be of significant consequence, but there is another way this can be approached in which the parcel starting P/T/Q is used to compute the LCL, and then the parcel value below the LCL has a constant virtual potential temperature. It would effectively preserve the values in the 1D arrays from the model. This would likely require significant modifications, though, since it looks like the LCL is an integer index rather than a computed value. My gut says "probably not", but I wanted to pose the question. 

3. Do these modifications need to be made to CALCAPE2? I'm not well versed in the UPP code and am not exactly sure what this does differently, and for what purpose. I do see some stuff about effective inflow layer calculations 0-3 km CAPE, which could be sensitive to this change as well.

Lastly, I noticed that in the MLCAPE code the pressure values are averaged. 
https://github.com/NOAA-EMC/UPP/blob/8f6caa9d04f1caf2ca8ce1b362aedcebf9b14c8c/sorc/ncep_post.fd/MISCLN.f#L3092-L3106
I could be misunderstanding what the code is doing (potentially related to my question about the T1D array), but canonically at SPC, the mixed-layer parcel is lifted from the shelter pressure value. The 100mb mean layer potential temperature is used as a proxy for boundary-layer mixing processes, but is still lifted from the surface. I wrote the code to maintain the existing functionality, so if we want to keep it that way we can... but also figured I'd mention the difference just in case. I'm also not sure what LB2 is and whether this average needs modification, too.

keltonhalbert committed:
Initial translation from Edward's email

keltonhalbert committed:
Shelter pressure should be included here too

BenjaminBlake-NOAA commented:
@keltonhalbert Thanks for working on this!  I'm tagging @jaymes-kenyon here as he may be more familiar with the UPP CAPE routines than I am, but here are my two cents:

1. It looks like the T1D and TBND arrays are actually for air temperature and not potential temperature.  T1D and TBND are used as inputs to the CALPOT routine, which calculates potential temperature.  CALPOT is called within MISCLN.f.

2. I am also thinking no, but it would be good to get confirmation on this.

3. I believe the same modifications should also be made to CALCAPE2 within UPP_PHYSICS.f.

4. I'm not sure what approach regarding the MLCAPE averaging is best, but perhaps we should stick with the current approach for now.

keltonhalbert commented:
>     1. It looks like the T1D and TBND arrays are actually for air temperature and not potential temperature.  T1D and TBND are used as inputs to the CALPOT routine, which calculates potential temperature.  CALPOT is called within MISCLN.f.

Ahhhh -- well that is definitely a difference worth noting. The averaging process being in T/P space instead of THETA space could result in some differences, particularly in models with greater grid spacing in the lowest levels. We can keep it in T/P space, but it would probably be more appropriate to average in potential temperature space. 

I can make the modifications to CALCAPE2 then!

EdwardColon-NOAA commented:
Hello Kelton and Ben,

Sorry for the late reply. The office desktop ethernet cable was being used
to address a laptop issue.

(1) I am in agreement with Ben regarding the T1D and TBND array units which
would complicate things a bit in the SCAPE and MUCAPE calculation. (2) The
direct substitution of T2M and Q2M is erroneous so, no, it needs to be
changed.(3) CALCAPE2 would need to reflect the changes in CALCAPE, and (4)
the MLCAPE averaging is simplistic and may require some weighting of
components but is the more important of the tasks.

Thank you for making the initial pull request, Kelton. Are you compiling
UPP on Jet?

-Edward

On Tue, Jul 22, 2025 at 2:35 PM Kelton Halbert ***@***.***>
wrote:

> *keltonhalbert* left a comment (NOAA-EMC/UPP#1275)
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3104278032>
>
> 1. It looks like the T1D and TBND arrays are actually for air temperature and not potential temperature.  T1D and TBND are used as inputs to the CALPOT routine, which calculates potential temperature.  CALPOT is called within MISCLN.f.
>
> Ahhhh -- well that is definitely a difference worth noting. The averaging
> process being in T/P space instead of THETA space could result in some
> differences, particularly in models with greater grid spacing in the lowest
> levels. We can keep it in T/P space, but it would probably be more
> appropriate to average in potential temperature space.
>
> I can make the modifications to CALCAPE2 then!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3104278032>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQFVIA7OH3CRFXMK4DT3J2AAFAVCNFSM6AAAAACCDPDYI2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMBUGI3TQMBTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

keltonhalbert committed:
removed calculation of potential temperature since it is in units of temperature, and added the same logic to CALCAPE2

keltonhalbert committed:
updated virtual mixed-layer calculation

keltonhalbert commented:
> (1) I am in agreement with Ben regarding the T1D and TBND array units which would complicate things a bit in the SCAPE and MUCAPE calculation. 
> (2) The direct substitution of T2M and Q2M is erroneous so, no, it needs to be changed. 
> (3) CALCAPE2 would need to reflect the changes in CALCAPE, and (4) the MLCAPE averaging is simplistic and may require some weighting of components but is the more important of the tasks. 

I have removed the calculations of potential temperature derived from the shelter fields, and have instead just done a direct read from the shelter arrays into the appropriate variable. Something that was missing was also the setting of the shelter pressure... the short variable names in FORTRAN make it hard for me to know for sure, but I think I set it to the correct variable `PKL` based on looking at the surrounding code. If someone could help confirm, that would be appreciated. I also copied there adjustments into `CALCAPE2`. 

A few additional questions and observations as I continue to familiarize myself with the code...

1. The shelter fields are currently only set when `ITYPE = 1` is set. I saw a comment that says for this type, P1D, T1D, and Q1D are "dummy arrays"... not entirely sure what this means here? The documentation says that this is for finding the maximum THETA-E layer, and it reads from P1D, T1D, etc. Either way, anything with this `ITYPE` should now include the contributions from the 2-meter fields. I think/hope...
2. When ITYPE=2, manual "intervention" is required. This is currently set up for mixed layer calculations, but there are others listed that may need to have this change included for consistency. See below, with line numbers corresponding to MISCLN.f.

- ITYPE = 1: (these are covered by changes to CALCAPE[2])
    - Most-unstable CAPE/CINH | L 3242
    - Significant Tornado Parameter (fixed-layer) | L4172

- ITYPE = 2: (These have to be covered by changes external to CALCAPE[2])
    - Best boundary layer CAPE/CINH | L2077
    - Mixed-layer CAPE/CINH | L3107
    - Effective Inflow layer | L3477
    - 0-3 km CAPE/CINH | L3602
    - mixed-layer LFC (part of STP calc -- explicitely sets virtual T??) | L4297

Something else I noticed is that in the call to CALCAPE for the virtual mixed-layer LFC (L4297), it explicitly makes the call to the virtual temperature correction for the mixed-layer, where the other routines do not. However, within CALCAPE, I see calls to `TVIRTUAL` as well.  My assumption was that CALCAPE is already doing the virtual temperature correction -- is this not the case? I'm a bit confused and could definitely use some clarification on what's happening there. 

I went ahead and updated the mixed-layer LFC calculation to include the 2-meter fields with the virtual temperature correction. However, it would be good to have some clarification on why the other calculations do not have or need these corrections. Mixed-layer, most-unstable, and surface-based parcel temperature definitions all need to use the virtual temperature consistently to be correct/accurate. 

>Are you compiling UPP on Jet?

Unfortunately I do not have access to Jet, and at one point I was given a WCOSS user account. However, that was earlier in the spring, and I never could get Tectia installed on my workstation due to staffing issues. Now, I am waiting for SPC to be given a license for the new software, SecureCRT, in order to access. So, in the mean time, I am using Spack to set up a local dev environment for UPP that I can at least test compilation on -- though I may have to employ assistance to test on actual output until I get on one of these systems.

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

EdwardColon-NOAA commented:
Hi Kelton,

-the shelter pressure is specified using PSHLTR. PKL is assigned the value
of PMID at the surface which seems to be the functional equivalent.

-for ITYPE=1, the variables P1D, T1D, and Q1D are not used directly in the
CAPE calculation although the arrays contain the pressure, temperature and
moisture of the air parcel lifted in the calculation of CAPE. So, I
suppose, they are referred to as "dummy arrays" in this context.

-There are a number of downstream field dependencies based on the CALCAPE
and CALCAPE2 calculations which you have listed for ITYPE=1 and ITYPE=2
categories as you have listed that would need to be updated for the sake of
consistency.

-Regarding LFC calculation, I believe you are correct. The virtual
temperature correction is being applied twice. That correction should be
removed outside of CALCAPE.

-Could you point me to the code block where you made the update to the
mixed-layer LFC calculation to include the 2-meter fields with the virtual
temperature correction?

-While waiting for WCOSS2 access, I can lend a hand in testing code updates
if you would like.

Thank you,

Edward



On Wed, Jul 23, 2025 at 4:02 PM Kelton Halbert ***@***.***>
wrote:

> *keltonhalbert* left a comment (NOAA-EMC/UPP#1275)
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3109964213>
>
> (1) I am in agreement with Ben regarding the T1D and TBND array units
> which would complicate things a bit in the SCAPE and MUCAPE calculation.
> (2) The direct substitution of T2M and Q2M is erroneous so, no, it needs
> to be changed.
> (3) CALCAPE2 would need to reflect the changes in CALCAPE, and (4) the
> MLCAPE averaging is simplistic and may require some weighting of components
> but is the more important of the tasks.
>
> I have removed the calculations of potential temperature derived from the
> shelter fields, and have instead just done a direct read from the shelter
> arrays into the appropriate variable. Something that was missing was also
> the setting of the shelter pressure... the short variable names in FORTRAN
> make it hard for me to know for sure, but I think I set it to the correct
> variable PKL based on looking at the surrounding code. If someone could
> help confirm, that would be appreciated. I also copied there adjustments
> into CALCAPE2.
>
> A few additional questions and observations as I continue to familiarize
> myself with the code...
>
>    1. The shelter fields are currently only set when ITYPE = 1 is set. I
>    saw a comment that says for this type, P1D, T1D, and Q1D are "dummy
>    arrays"... not entirely sure what this means here? The documentation says
>    that this is for finding the maximum THETA-E layer, and it reads from P1D,
>    T1D, etc. Either way, anything with this ITYPE should now include the
>    contributions from the 2-meter fields. I think/hope...
>    2. When ITYPE=2, manual "intervention" is required. This is currently
>    set up for mixed layer calculations, but there are others listed that may
>    need to have this change included for consistency. See below, with line
>    numbers corresponding to MISCLN.f.
>
>
>    -
>
>    ITYPE = 1: (these are covered by changes to CALCAPE[2])
>    - Most-unstable CAPE/CINH | L 3242
>       - Significant Tornado Parameter (fixed-layer) | L4172
>    -
>
>    ITYPE = 2: (These have to be covered by changes external to CALCAPE[2])
>    - Best boundary layer CAPE/CINH | L2077
>       - Mixed-layer CAPE/CINH | L3107
>       - Effective Inflow layer | L3477
>       - 0-3 km CAPE/CINH | L3602
>       - mixed-layer LFC (part of STP calc -- explicitely sets virtual
>       T??) | L4297
>
> Something else I noticed is that in the call to CALCAPE for the virtual
> mixed-layer LFC (L4297), it explicitly makes the call to the virtual
> temperature correction for the mixed-layer, where the other routines do
> not. However, within CALCAPE, I see calls to TVIRTUAL as well. My
> assumption was that CALCAPE is already doing the virtual temperature
> correction -- is this not the case? I'm a bit confused and could definitely
> use some clarification on what's happening there.
>
> I went ahead and updated the mixed-layer LFC calculation to include the
> 2-meter fields with the virtual temperature correction. However, it would
> be good to have some clarification on why the other calculations do not
> have or need these corrections. Mixed-layer, most-unstable, and
> surface-based parcel temperature definitions all need to use the virtual
> temperature consistently to be correct/accurate.
>
> Are you compiling UPP on Jet?
>
> Unfortunately I do not have access to Jet, and at one point I was given a
> WCOSS user account. However, that was earlier in the spring, and I never
> could get Tectia installed on my workstation due to staffing issues. Now, I
> am waiting for SPC to be given a license for the new software, SecureCRT,
> in order to access. So, in the mean time, I am using Spack to set up a
> local dev environment for UPP that I can at least test compilation on --
> though I may have to employ assistance to test on actual output until I get
> on one of these systems.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3109964213>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQCZBX3LCCYKDU4JNMT3J7SYVAVCNFSM6AAAAACCDPDYI2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMBZHE3DIMRRGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

keltonhalbert commented:
Hi @EdwardColon-NOAA, 

Thank you for some of the clarifications, and your willingness to be an intermediate compiler/tester!

Regarding your question about the 2-meter virtual temperature fields, this is the static link to the lines in MISCLN.f: https://github.com/NOAA-EMC/UPP/blob/0ac3cfaa73907f2bc5c2a7d7f2237335eee9d0ba/sorc/ncep_post.fd/MISCLN.f#L4285-L4292

This is the section that is potentially "double correcting" for the virtual temperature when computing the mixed-layer LFC. It appears to then call CALCAPE2 with this configuration.

keltonhalbert commented:
If we are in agreement on this, I have removed the calls to `TVIRTUAL` outside of CALCAPE[2]. It appears this was the only instance of that function being called in MISCLN.f, so I removed it from the import at the top of the file. 

The following calls to `CALCAPE[2]` have been updated to include the 2-meter values:

- Mixed-layer CAPE/CINH | L3107
- 0-3 km CAPE/CINH | L3602
- mixed-layer LFC (part of STP calc -- no longer sets virtual T) | L4297

The two remaining calculations that need addressing: 
 - Best boundary layer CAPE/CINH | L2077
 - Effective Inflow layer | L3477


**Best Boundary Layer CAPE**

https://github.com/keltonhalbert/UPP/blob/581d81acdbeac1e08b1fa1d33e85663476823944/sorc/ncep_post.fd/MISCLN.f#L2059-L2078

**Effective Inflow Layer**

https://github.com/keltonhalbert/UPP/blob/581d81acdbeac1e08b1fa1d33e85663476823944/sorc/ncep_post.fd/MISCLN.f#L3459-L3478

Both of these are loops over a vertical coordinate, with the best boundary layer being over `NBND` and the effective inflow layer being over `LM`. Neither of these pose easy ability to add the shelter fields to the loop indices, but there are probably a few options. I'm having a hard time deciphering the "best boundary layer CAPE", but I am assuming it is taking the maximum value... so there should be a way to compare the shelter field CAPE with the model levels and take the maximum. The effective inflow layer would be a bit more tricky, since it is also using the equilibrium level value based on the effective inflow criteria.

keltonhalbert committed:
removed calls to TVIRTUAL that were not necessary, added shelter fields to additional CAPE calculations

keltonhalbert committed:
Merge branch 'cape-cinh-2m' of github.com:keltonhalbert/UPP into cape-cinh-2m

keltonhalbert committed:
Merge branch 'develop' of github.com:NOAA-EMC/UPP into cape-cinh-2m

keltonhalbert committed:
fixed missing paren

keltonhalbert commented:
After many trials and tribulations, I finally managed to get the UPP and its dependencies compiled and installed. Turned out, the version of CRTM spack defaults to does not work, but setting it to the version in the spack.yaml file for CI worked. 

I can confirm that my changes at least compile. I can't speak to the quality of the output, but it at least builds.

EdwardColon-NOAA commented:
Hi Kelton,

I'll clone your branch and run a test.

Thanks,

Edward

On Mon, Jul 28, 2025 at 6:30 PM Kelton Halbert ***@***.***>
wrote:

> *keltonhalbert* left a comment (NOAA-EMC/UPP#1275)
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3130041812>
>
> After many trials and tribulations, I finally managed to get the UPP and
> its dependencies compiled and installed. Turned out, the version of CRTM
> spack defaults to does not work, but setting it to the version in the
> spack.yaml file for CI worked.
>
> I can confirm that my changes at least compile. I can't speak to the
> quality of the output, but it at least builds.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3130041812>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQGTMRZ7FIBILIDLV6L3K2QA5AVCNFSM6AAAAACCDPDYI2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMZQGA2DCOBRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA commented:
Hi Kelton,

The merging of your branch into the develop branch caused some issues in
testing as I have set up the regression test with reference files derived
from the older version of UPP that we are using in the 3DRTMA. I integrated
the changes that you made in your branch into the UPP branch that we are
currently using (https://github.com/EdwardColon-NOAA/UPP/tree/release/3drtma)
and ran a regression test. It passed for both the CONUS and Alaska domains.
Before I perform any further testing, I need to ensure my fork is
compatible with the release/3drtma repository. An initial comparison
suggests differences.

Ultimately, we will need to include your changes in the develop branch but,
for now, initial testing would be easier on the older branch.

Thanks,

Edward



On Mon, Jul 28, 2025 at 6:50 PM Edward Colon - NOAA Affiliate <
***@***.***> wrote:

> Hi Kelton,
>
> I'll clone your branch and run a test.
>
> Thanks,
>
> Edward
>
> On Mon, Jul 28, 2025 at 6:30 PM Kelton Halbert ***@***.***>
> wrote:
>
>> *keltonhalbert* left a comment (NOAA-EMC/UPP#1275)
>> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3130041812>
>>
>> After many trials and tribulations, I finally managed to get the UPP and
>> its dependencies compiled and installed. Turned out, the version of CRTM
>> spack defaults to does not work, but setting it to the version in the
>> spack.yaml file for CI worked.
>>
>> I can confirm that my changes at least compile. I can't speak to the
>> quality of the output, but it at least builds.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3130041812>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACFXXQGTMRZ7FIBILIDLV6L3K2QA5AVCNFSM6AAAAACCDPDYI2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMZQGA2DCOBRGI>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> -------------------------------------------------
> Edward Colón
> *Lynker* at NOAA/NWS/NCEP/EMC
> 5830 University Research Ct.
> Rm #2025
> Riverdale, MD 20737
> Office: (301) 683-3815
> Cell: (301) 213-3566
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

keltonhalbert commented:
@EdwardColon-NOAA, 

Since the changes are relatively lightweight, I can apply them to a fork of `release/3drtma` if need be. The last thing that needs to be figured out is how to include the shelter fields in the vertical loops for the effective inflow layer and best boundary layer CAPE. I  have been working forecast shifts this week so I haven't come up with a solution yet, but if and when that is the case, can work to apply it to the other branch as well.

EdwardColon-NOAA commented:
Thanks Kelton,

I'll go ahead and finish the release/3drtma update today.

-Edward

On Wed, Jul 30, 2025 at 9:59 AM Kelton Halbert ***@***.***>
wrote:

> *keltonhalbert* left a comment (NOAA-EMC/UPP#1275)
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3136480102>
>
> @EdwardColon-NOAA <https://github.com/EdwardColon-NOAA>,
>
> Since the changes are relatively lightweight, I can apply them to a fork
> of release/3drtma if need be. The last thing that needs to be figured out
> is how to include the shelter fields in the vertical loops for the
> effective inflow layer and best boundary layer CAPE. I have been working
> forecast shifts this week so I haven't come up with a solution yet, but if
> and when that is the case, can work to apply it to the other branch as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/UPP/pull/1275#issuecomment-3136480102>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQGCM5GK5IB7QRRYFZ33LDFTRAVCNFSM6AAAAACCDPDYI2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMZWGQ4DAMJQGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

BenjaminBlake-NOAA commented:
@keltonhalbert Work is in progress on updating the 3DRTMA UPP.  After this work is complete we will revisit your CAPE/CIN PR.  Thanks for your patience!

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

keltonhalbert commented:
> @keltonhalbert Work is in progress on updating the 3DRTMA UPP. After this work is complete we will revisit your CAPE/CIN PR. Thanks for your patience!

No worries @BenjaminBlake-NOAA! I'm in no rush over here and have plenty to occupy myself with until it's time to revisit this.

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

BenjaminBlake-NOAA commented:
Hi @keltonhalbert , I just opened a draft PR (#1372 ) to the release/3drtma_v1 branch containing your changes from this PR.  We would like to target the 3DRTMA branch first for these CAPE/CINH changes.  When you get a chance, let me know if I have accurately captured your UPP code changes there.  Thanks!

BenjaminBlake-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

Commit:
Add changes from PR to 3DRTMA release branch

BenjaminBlake-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -3096,9 +3098,27 @@ SUBROUTINE MISCLN
+                   P1D(I,J)  = (PBND(I,J,1) + PBND(I,J,2) + PBND(I,J,3) + &
+                                PSHLTR(I,J))/4
+                   T1D(I,J)  = (TBND(I,J,1) + TBND(I,J,2) + TBND(I,J,3) + &
+                                TSHLTR(I,J))/4
```
Thanks @GangZhao-NOAA for catching a glitch:   TSHLTR inside UPP is potential temperature. I suggest changing it into 2-m T as follows:
```
TSHLTR(I,J) --> TSHLTR(I,J)*(PSHLTR(I,J)/P1000)**CAPA
```
Tagging @GangZhao-NOAA @BenjaminBlake-NOAA @keltonhalbert

BenjaminBlake-NOAA replied:
@GangZhao-NOAA Thanks for catching this - @WenMeng-NOAA I'll get this updated tomorrow.  At some point we'll also want to get this updated in the 3DRTMA release branch, perhaps combined with a future PR that Gang will submit.

BenjaminBlake-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

BenjaminBlake-NOAA commented on sorc/ncep_post.fd/UPP_PHYSICS.f:
```diff
@@ -694,8 +696,14 @@ SUBROUTINE CALCAPE(ITYPE,DPBND,P1D,T1D,Q1D,L1D,CAPE,    &
-                  QBTK   = max(0.0, Q(I,J,KB))
+                  IF (capecin_2m .AND. KB == LM) THEN 
+                      PKL = PSHLTR(I,J)
+                      TBTK = TSHLTR(I,J)
```
Should TSHLTR in UPP_PHYSICS.f also be updated to TSHLTR*(PSHLTR/P1000)^CAPA ?

WenMeng-NOAA replied:
@BenjaminBlake-NOAA Yes, it would be good to apply to ALL "TSHLTR" related to CAPE updates.

Commit:
Update TSHLTR in CAPE/CIN calculation to 2-m temperature

WenMeng-NOAA commented:
@BenjaminBlake-NOAA There are a syntax errors as below, which might cause CI failure:
'''
Error: Symbol ‘p1000’ at (1) has no IMPLICIT type; did you mean ‘p00k’?
/home/runner/work/UPP/UPP/UPP/sorc/ncep_post.fd/UPP_PHYSICS.f:701:59:
Error: Symbol ‘p1000’ at (1) has no IMPLICIT type; did you mean ‘p00k’?
make[2]: *** [sorc/ncep_post.fd/CMakeFiles/upp.dir/build.make:1625: sorc/ncep_post.fd/CMakeFiles/upp.dir/UPP_PHYSICS.f.o] Error 1
'''

BenjaminBlake-NOAA commented:
@WenMeng-NOAA Thanks, looks like I need to declare P1000 - I'll get this fixed.

Commit:
Declare P1000

BenjaminBlake-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA commented:
The baseline of gfs test has the following changes:
```
892:2486066395:CAPE:surface:rpn_corr=0.973987:rpn_rms=245.415
893:2488576822:CIN:surface:rpn_corr=0.744611:rpn_rms=37.2207
1044:3071206478:CAPE:90-0 mb above ground:rpn_corr=0.987647:rpn_rms=94.6981
1045:3073362134:CIN:90-0 mb above ground:rpn_corr=0.741131:rpn_rms=33.1882
1046:3074565713:CAPE:255-0 mb above ground:rpn_corr=0.972574:rpn_rms=245.248
1047:3076869436:CIN:255-0 mb above ground:rpn_corr=0.687106:rpn_rms=35.5124
1048:3077818705:PLPL:255-0 mb above ground:rpn_corr=0.969716:rpn_rms=4173
```

Commit:
Add capecin_2m namelist option to affected model applications

BenjaminBlake-NOAA commented:
@keltonhalbert After discussing with @WenMeng-NOAA , we are going to move forward with merging the work in this PR.  For most affected applications (except 3DRTMA) we will leave capecin_2m=.false. for now until further testing/evaluation has been completed.  I know there is a desire to add these changes to GFS and RRFS.  Thank you for your significant contribution to UPP!

@WenMeng-NOAA With the current settings, there are only baseline changes to the 3DRTMA test (capecin_2m=.true.).  There are baseline changes for RRFS but they can be ignored because I know you already updated the baselines for my RRFS PR, which we can process first.  GEFSv12 and RRFS_IFI_missing were the only RTs that did not have baseline changes when capecin_2m=.true.

WenMeng-NOAA commented on sorc/ncep_post.fd/CTLBLK.f:
```diff
@@ -78,6 +79,7 @@ module CTLBLK_mod
   logical :: slrutah_on    !< Calculate snow to liquid ratio (SLR) using method from University of Utah.
   logical :: gtg_on        !< Turn on GTG (Graphical Turbulence Guidance)
   logical :: method_blsn   !< Turn on blowing snow effect on visibility diagnostic
+  logical :: capecin_2m    !< TUrn on option to calculate CAPE and CIN using 2-m fields
```
@BenjaminBlake-NOAA Could you initialize capecin_2m as false in case users want to upgrade upp hash in their experimental UFS versions.

BenjaminBlake-NOAA replied:
@WenMeng-NOAA I removed capecin_2m from the RT scripts since the default option is false.

It looks like capecin_2m is initialized to false (default value) in WRFPOST.F.  Do I also need to initialize capecin_2m = .false. in CTLBLK.f?

WenMeng-NOAA replied:
@BenjaminBlake-NOAA Yes. The initialization in WRFPOST.F only works for off-line post, not for in-line post at UFS side.

Commit:
Remove capecin_2m from UPP namelist - default option is false

Commit:
Initialize capecin_2m as false in CTLBLK.f

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

BenjaminBlake-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

BenjaminBlake-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

BenjaminBlake-NOAA commented:
The CAPE/CIN modifications in this PR result in changes to the following UPP RTs when the capecin_2m namelist option is enabled (all tests except gefsv12 and rrfs_ifi_missing):
3drtma, aqm, dafs, gcafs, gefsv13, gfs, hafs, hrrr, hrrr_ifi, mpas, mpas_hfip, nmmb, rap, rrfs, rrfs_ifi, sfs

The variables that are changed with each test are outlined in this [Google Doc](https://docs.google.com/document/d/10n5w1hmCD9ym9hXiByTk65NyImbxaWk7NOW_qcHoIsw/edit?usp=sharing) (accessible to anyone with a NOAA account)

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA committed:
Merge branch 'develop' into cape-cinh-2m

WenMeng-NOAA commented:
There are baseline changes in 3drtma test with the flag 'capecin_2m' disabled:
```
WRFTWO.GrbF00.diff
167:137685110:MLFC:level of free convection:rpn_corr=0.878021:rpn_rms=809.152
172:146090030:ESP:0-3000 m above ground:rpn_corr=0.987917:rpn_rms=4.8645
```
@BenjaminBlake-NOAA We need to identify the above changes.

WenMeng-NOAA commented:
> There are baseline changes in 3drtma test with the flag 'capecin_2m' disabled:
> 
> ```
> WRFTWO.GrbF00.diff
> 167:137685110:MLFC:level of free convection:rpn_corr=0.878021:rpn_rms=809.152
> 172:146090030:ESP:0-3000 m above ground:rpn_corr=0.987917:rpn_rms=4.8645
> ```
> 
> @BenjaminBlake-NOAA We need to identify the above changes.

There are a bug fix that replaces 'TVIRTUAL' with 'TBND' in MISCLN.f. So the above changes are expected.

BenjaminBlake-NOAA commented:
> There are a bug fix that replaces 'TVIRTUAL' with 'TBND' in MISCLN.f. So the above changes are expected.

@WenMeng-NOAA Yes that's correct, the bug fix for TVIRTUAL should result in these changes.

WenMeng-NOAA committed:
Add RT log for WCOSS2

BenjaminBlake-NOAA commented:
@clyden-noaa This PR is ready for the RTs on the RDHPCs.  There are baseline changes expected for 3DRTMA due to a bug fix related to the virtual temperature correction.  I've generated 3DRTMA baseline data for you on Ursa:
```
/scratch4/NCEPDEV/stmp/Benjamin.Blake/upp-URSA-intel/3drtma_2025091010/WRFTWO.GrbF00
/scratch4/NCEPDEV/stmp/Benjamin.Blake/upp-URSA-intelllvm/3drtma_2025091010/WRFTWO.GrbF00
```
Thanks!

GangZhao-NOAA reviewed (approved):
The modifications look good to me.

Thanks!

MatthewMorris-NOAA reviewed (approved):
Looks good to me!

clyden-noaa committed:
Ursa RT log for intel compiler

clyden-noaa committed:
Ursa RT log for intelllvm compiler

clyden-noaa committed:
Orion RT log for intel compiler

clyden-noaa committed:
Hercules RT log for intel compiler

clyden-noaa reviewed (commented):
@WenMeng-NOAA @BenjaminBlake-NOAA @keltonhalbert 
All testing has been completed
All baseline files have been copied into place on all RDHPCs
All RTs ran successfully without errors and within established timing thresholds on all RDHPCs

**Ursa: intel compiler**
```
Check tests:
['sfs', 'aqm', 'gefsv12', 'gefsv13', 'nmmb', 'rap', 'hrrr', 'hafs', '3drtma', 'mpas', 'mpas_hfip', 'gcafs', 'rrfs', 'rrfs_ifi_missing', 'gfs']
No changes in test results detected.
```

**Ursa: intelllvm compiler**
```
Check tests:
['sfs', 'aqm', 'gefsv12', 'gefsv13', 'nmmb', 'rap', 'hrrr', 'hafs', '3drtma', 'mpas', 'mpas_hfip', 'gcafs', 'rrfs', 'rrfs_ifi_missing', 'gfs']
No changes in test results detected.
```

**Orion: intel compiler**
```
Check tests:
['sfs', 'aqm', 'gefsv12', 'gefsv13', 'nmmb', 'rap', 'hrrr', 'hafs', '3drtma', 'mpas', 'mpas_hfip', 'gcafs', 'rrfs', 'rrfs_ifi_missing', 'gfs']
No changes in test results detected.
```

**Hercules: intel compiler**
```
Check tests:
['sfs', 'aqm', 'gefsv12', 'gefsv13', 'nmmb', 'rap', 'hrrr', 'hafs', '3drtma', 'mpas', 'mpas_hfip', 'gcafs', 'rrfs', 'rrfs_ifi_missing', 'gfs']
No changes in test results detected.
```

clyden-noaa reviewed (approved):
Approved

BenjaminBlake-NOAA reviewed (approved):
This PR is ready to be merged.  Thanks for all of the work/collaboration on this everyone!
